packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,315 @@
+
+        this.tokenEmitter.emit('access_token', { token })
+
+        // Reset our failure count
+        this.retryCount = 0
+
+        // Refresh the token after half the expiry time passes
+        timeUntilRefreshInMs = (token.expires_in / 2) * 1000
+      } catch (err) {
+        // Something went really wrong with the body, lets surface an error and try again?
+        this.tokenEmitter.emit('access_token', { error: err })
+        this.retryCount = 0 // Reset because we've emitted an error
+
+        timeUntilRefreshInMs = backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+      }
+    } else if (response.status === 429) {
+      this.retryCount++
+      this.lastError = `[${response.status}] ${response.statusText}`
+      if (headers && headers['x-ratelimit-reset']) {
+        const rateLimitResetTimestamp = parseInt(
+          headers['x-ratelimit-reset'],
+          10
+        )
+        if (isFinite(rateLimitResetTimestamp)) {
+          timeUntilRefreshInMs =
+            rateLimitResetTimestamp -
+            Date.now() +
+            this.clockSkewInSeconds * 1000
+        } else {
+          timeUntilRefreshInMs = 5 * 1000
+        }
+        // We want subsequent calls to get_token to be able to interrupt our
+        //  Timeout when it's waiting for e.g. a long normal expiration, but
+        //  not when we're waiting for a rate limit reset. Sleep instead.
+        await sleep(timeUntilRefreshInMs)
+        timeUntilRefreshInMs = 0
+      }
+    } else if ([400, 401, 415].includes(response.status)) {
+      // Unrecoverable errors
+      this.retryCount = 0
+      this.tokenEmitter.emit('access_token', {
+        error: new Error(`[${response.status}] ${response.statusText}`),
+      })
+      this.stopPoller()
+      return
+    } else {
+      this.retryCount++
+      this.lastError = new Error(`[${response.status}] ${response.statusText}`)
+      timeUntilRefreshInMs = backoff({
+        attempt: this.retryCount,
+        minTimeout: 25,
+        maxTimeout: 1000,
+      })
+    }
+
+    if (this.retryCount % this.maxRetries === 0) {
+      this.tokenEmitter.emit('access_token', { error: this.lastError })
+      // TODO: figure out timing and whether to reset retries?
+    }
+    this.pollerTimer = setTimeout(
+      this.pollerLoop.bind(this),
+      timeUntilRefreshInMs
+    )?.unref()
+  }
+
+  stopPoller() {
+    clearTimeout(this.pollerTimer)
+  }
+
+  /**
+   * Solely responsible for building the HTTP request and calling the token service.
+   */
+  private requestAccessToken(): Promise<HTTPResponse> {
+    const jti = uuid()
+    const currentUTCInSeconds =
+      Math.round(Date.now() / 1000) - this.clockSkewInSeconds
+    const jwtBody = {
+      iss: this.clientId,
+      sub: this.clientId,
+      aud: this.authServer,
+      iat: currentUTCInSeconds - 5,
+      exp: currentUTCInSeconds + 55,
```

Comment:
what are these magic numbers for? can we put them into constants with a better explanation for non-well-versed oauth people like me not to screw these up ?

Reply:
No more magic numbers!